packages/kbn-content-management-utils/src/saved_object_content_storage.ts

```diff
@@ -0,0 +1,391 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Boom from '@hapi/boom';
+import type { SearchQuery, SearchIn } from '@kbn/content-management-plugin/common';
+import type {
+  ContentStorage,
+  StorageContext,
+  MSearchConfig,
+} from '@kbn/content-management-plugin/server';
+import type {
+  SavedObject,
+  SavedObjectReference,
+  SavedObjectsFindOptions,
+  SavedObjectsCreateOptions,
+  SavedObjectsUpdateOptions,
+  SavedObjectsFindResult,
+} from '@kbn/core-saved-objects-api-server';
+import type {
+  CMCrudTypes,
+  ServicesDefinitionSet,
+  SOWithMetadata,
+  SOWithMetadataPartial,
+} from './types';
+
+const savedObjectClientFromRequest = async (ctx: StorageContext) => {
+  if (!ctx.requestHandlerContext) {
+    throw new Error('Storage context.requestHandlerContext missing.');
+  }
+
+  const { savedObjects } = await ctx.requestHandlerContext.core;
+  return savedObjects.client;
+};
+
+type PartialSavedObject<T> = Omit<SavedObject<Partial<T>>, 'references'> & {
+  references: SavedObjectReference[] | undefined;
+};
+
+function savedObjectToItem<Attributes extends object, Item extends SOWithMetadata>(
+  savedObject: SavedObject<Attributes>,
+  partial: false
+): Item;
+
+function savedObjectToItem<Attributes extends object, PartialItem extends SOWithMetadata>(
+  savedObject: PartialSavedObject<Attributes>,
+  partial: true
+): PartialItem;
+
+function savedObjectToItem<Attributes extends object>(
+  savedObject: SavedObject<Attributes> | PartialSavedObject<Attributes>
+): SOWithMetadata | SOWithMetadataPartial {
+  const {
+    id,
+    type,
+    updated_at: updatedAt,
+    created_at: createdAt,
+    attributes,
+    references,
+    error,
+    namespaces,
+  } = savedObject;
+
+  return {
+    id,
+    type,
+    updatedAt,
+    createdAt,
+    attributes,
+    references,
+    error,
+    namespaces,
+  };
+}
+
+export interface SearchArgsToSOFindOptionsOptionsDefault {
+  fields?: string[];
+  searchFields?: string[];
```

Comment:
I don't think we want the UI to be able to search on _any_ field in the DB or to ask for any fields to be returned. I think we want the options to be constraint. Thoughts @jloleysens ?

Reply:
>I think we want the options to be constraint.

Why do we want the options to be constrained? As things are, we have the potential to implement constraints but are they necessary?

Reply:
If "fieldA" is changed to "fieldB" and we allow _any_ field to be queried from the UI then we won't be able to change the request and rename "fieldA" to "fieldB" before hitting the search layer.

Reply:
The concern there is that the request is for the literal field rather than the content. There's nothing prohibiting the request for "fieldA" to query the SO "fieldB" at some later date. This could be handled by the existing transform code.

Reply:
Browser is on v1, it only knows about "fieldA", it queries to get result... but server is on v2 and only knows about "fieldB".. it returns empty results...
The UI is "broken", it _thinks_ there are no result.

Maybe the @elastic/kibana-core can shed some light if it is actually OK to fetch any fields from the UI.

Reply:
@sebelga In that case the v2 server would have the transform to convert "fieldA" to "fieldB".

Reply:
Unless I am missing something... your code _is_ the server 😊

Reply:
I will mark this as resolved although from what I understood from the @elastic/kibana-core team we should not expose DB fields directly to the UI but be explicit about which fields can be searched. I let them chime in if they share the concern.